### PR TITLE
[bugfix] Favor + fp16

### LINKED ISSUE #96: Favor breaks with fp16
# 🐛 Bug
The loss nans immediately

## Command
`python3 microGPT.py` when precision is set to 16

## Environment
Unrelated

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Fixes #96
There was no bug per say, the Softmax approximation for Favor just requires a big dynamic range and the attention computation would overflow in fp16. I tested a couple of takes (limit the values a posteriori, catch inf etc..) but this would be ugly and break the backward graph, in the end keeping this nugget of computations in fp32 is the safest bet I think. Note that we only promote fp16 tensors, not bfloat16 for instance

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
cc @xwhan, probably too late for you but FYI works fine now

File: examples/microGPT.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def __init__(
         vocab_size,
         weight_decay=0.1,
         betas=(0.9, 0.95),
-        learning_rate=6e-4,
+        learning_rate=1e-4,
```
Comment (blefaudeux) [inline examples/microGPT.py:31]: before this PR favor cannot run in fp16 (the very first FW dies). With this PR it runs just fine, but it requires a LR a little lower than other mechanisms to prevent overflow later on

**Review by dianaml0 — APPROVED** [review]
Interesting bug, LGTM!

**Comment by blefaudeux** [comment]
> Interesting bug, LGTM!

yeah it was really perplexing, k_prime and q_prime are completely fine, then compute att and bam, NaNs and inf all over the place... That's where the fp16 dynamic bites, 65k is not a lot before overflowing. I tried to post fix that in the beginning (knowing that it overflew but that the actual result was fine out of principle), but turns out that it's not completely trivial, not very clean, and not differentiable, so fp32 looked like the safest bet
